File: 2900/go/2900.go
func getLongestSubsequence(words []string, groups []int) []string {
    n := len(words)
    // dp, preWord
    dp := make([]int, n)
    pre := make([]int, n)

    for i := 0; i < n; i++{
        dp[i] = 1
    }

    // state transition
    for i := 1; i < n; i++ {
        for j := 0; j < i; j++{
            if groups[i] != groups[j] {
                if dp[i] < dp[j] + 1 {
                    dp[i] = dp[j] + 1
                    pre[i] = j
                }
            }
        }
    }

    maxL := -1
    endId := 0
    for i := 0; i < n; i++ {
        if maxL < dp[i] {
            maxL = dp[i]
            endId = i
        }
    }

    result := make([]string, maxL)
    for i := maxL-1; i >= 0; i-- {
        result[i] = words[endId]
        endId = pre[endId]
    }

    return result
}

File: 2900/js/2900.js
/**
 * @param {string[]} words
 * @param {number[]} groups
 * @return {string[]}
 */
var getLongestSubsequence = function(words, groups) {
    let n = words.length;

    // make dp, preList
    let dp = new Array(n).fill(1);
    let preL = new Array(n).fill(-1);

    // state transition
    for (let i = 0; i < n; i++){
        for (let j = 0; j < i; j++){
            if (groups[i] != groups[j]){
                if (dp[i] < dp[j]+1){
                    dp[i] = dp[j] + 1;
                    preL[i] = j;
                }
            }
        }
    }

    // find max
    let maxL = 0, endId = 0;
    for (let i = 0; i < n; i++){
        if (maxL < dp[i]){
            maxL = dp[i];
            endId = i;
        }
    }

    let result = new Array("").fill(maxL);
    for (let i = maxL-1; i >= 0; i--){
        result[i] = words[endId];
        endId = preL[endId];
    }

    return result;
};